### ISSUE TITLE
error with dsh shm deploy, azureclicrdential

### ISSUE THREAD

**Issue by mattwestby** [alan-turing-institute/data-safe-haven#2184]
When using the command dsh shm deploy to initially start deployment I receive the following errors after authentication with az login;

Are these details correct? [y/n] (y): y
Please authenticate with Azure: run 'az login' using infrastructure administrator credentials.
Error getting account information from Azure CLI.
Could not load list of groups.

I have owner permissions on the subscription I'm attempting to deploy into and when i check my az token after az login it looks ok. Not sure what the problem is?

**Comment by craddm** [comment]
Hi Matt,

`Could not load list of groups` is generated during interactions with Graph API, so that suggests to me the issue may not be with your Azure credentials per se - so not the infrastructure admin account you log in to using `az login` - but with your Entra ID administrator credentials. For us, we have one account that and a separate global administrator account on the Entra ID tenant that'll be associated with the SRE/SHM.

It's possible that the path to producing the error you're receiving is not quite correct and is not quite giving you enough information to clarify the source of the problem - I'll look into that.

Cheers,
Matt

**Comment by mattwestby** [comment]
Hi Matt,

Ah ok we do have a separate Entra ID to manage the SRE users but I've not authenticated with that during this stage? Did I need to authenticate with the global admin in that entra as part of this stage?
Thanks
Matt

**Comment by craddm** [comment]
I haven't been able to replicate your problem as yet, but it does seem to happen before you need to be authenticated with the Graph API, contrary to what I suggested above. So I think we can forget that. 

Where this error seems to come up is when retrieving the admin group id from the admin group name you provide in the `context`. I would check that the Azure account you are logged in with is a member of an appropriate security group and has sufficient permissions ~~(at least `Contributor`)~~(sorry, reread your original report and can see you have these!), and that the name of the security group is correct

**Comment by mattwestby** [comment]
Hi Matt,

So the admin group I provided in the context is a admin group that is in the “SRE” entra ID and not the same entra as the infra deployment. I have owner on the infra deployment subscription. Is that --admin-group-name meant to be in the same entra as the infra deployment tenant?

Thanks

Matt

**Comment by craddm** [comment]
> Hi Matt, So the admin group I provided in the context is a admin group that is in the “SRE” entra ID and not the same entra as the infra deployment. I have owner on the infra deployment subscription. Is that --admin-group-name meant to be in the same entra as the infra deployment tenant? Thanks Matt 

Hi Matt,

Yes, the admin group name should be in the infra deployment tenant Entra - we assign roles to the groups and users to the groups rather than directly assigning roles to users. 

I think the documentation could definitely be clearer on this

Cheers,
Matt

**Comment by mattwestby** [comment]
HI Matt,

Thanks, for this what permissions are then required in the infra entra?  I still get the error even though I’ve specified a group name in that entra.

And this switch on the dsh shm deploy command

“--entra-tenant-id”

This is tenant id for the “SRE” users entra?

Thanks

Matt

**Comment by craddm** [comment]
> HI Matt, Thanks, for this what permissions are then required in the infra entra? I still get the error even though I’ve specified a group name in that entra. 

Our admin group is set up with the `Owner` and `Billing Reader` roles by default; `Contributor` _should_ be enough. And then it's just our admins that are members of that group. 

> And this switch on the dsh shm deploy command “--entra-tenant-id” This is tenant id for the “SRE” users entra? Thanks Matt 

Yes, that's the SRE users Entra ID

Cheers,
Matt

**Comment by mattwestby** [comment]
HI Matt,

I’m still not quite clear on what the entra group requires, are you saying owner and billing reader / contributor on the subscription where the SHM will be deployed into?

Thanks

Matt

**Comment by craddm** [comment]
Apologies for lack of clarity. It *should* be sufficient to give the admin group (and note that this is a security group) the Contributor role. If that fails, try giving them Owner. We know that the admin group having Owner is enough on our end, so if that still fails, the problem is something else. I just threw in that ours also has Billing Reader as it's another role that our group has; it shouldn't be necessary, but perhaps if you want to fully replicate the roles our group has then it's handy to know it has that too. And yes these roles are for the subscription that the SRE will be deployed into.

**Comment by mattwestby** [comment]
Hi Matt

I assigned the owner role for that group on the subscription level but I still get this error:

***@***.***

So its something with the auth?

Thanks

Matt

**Comment by craddm** [comment]
Hi Matt,

The error seems to be missing from your last message. Is it the same error as before, about the groups?

**Comment by mattwestby** [comment]
Hi Matt,

Yea,

***@***.***

Thanks
Matt

**Comment by mattwestby** [comment]
Hi Matt

I’ve resolved it! I had to login to Azure CLI with

az login --scope https://graph.microsoft.com/.default

this allowed me to get the token correctly.

Thanks

Matt

From: Matthew Westby (staff)
Sent: 10 September 2024 13:58
To: alan-turing-institute/data-safe-haven ***@***.***>; alan-turing-institute/data-safe-haven ***@***.***>
Cc: Manual ***@***.***>
Subject: RE: [alan-turing-institute/data-safe-haven] error with dsh shm deploy, azureclicrdential (Issue #2184)

Hi Matt,

Yea,

***@***.***

Thanks
Matt

**Comment by craddm** [comment]
Some additional information from Matt by email. Matt was not being prompted for MFA when using `az login`, and was unable to acquire an access token. 

This was manifesting as an error during SHM deployment at the step where it attempts to match the admin group name provided by the user to its respective group id on Azure. 

I've been unable to replicate these problems locally. I tried giving incorrect admin group names and/or incorrect tenant ids in the configs, but nothing doing. I tried running the deployment from Windows, not problems observed.

For Matt, attempting to get an access token through the Azure CLI yielded the prompt to login using `az login --scope https://graph.microsoft.com/.default`, and that allowed him to successfully get an access token with sufficient scope over the graph API.

I'm wondering if there is some kind of default scope applied to the login that can be set in the portal through a policy of some sort, and whether that differs between our tenant and Matt's. I can't check this out as I do not have sufficient rights on our dev subscription to investigate how its Entra ID tenant is set up.

**Comment by mattwestby** [comment]
Thanks Matt,

I just wanted to check another point with the deployment of the SHM, ive run the dsh shm deploy and its passed the point of checking the domain verification;

***@***.***

And its deployed the following;

***@***.***

Is that the SHM deployed as it doesn’t produce any further output after verifying the domain?

Thanks
Matt

**Comment by craddm** [comment]
Hi Matt,

Yes, the last two lines of the output from deploying the SHM should be something like

```
Ensured that DNS TXT record @ exists in zone green.develop.turingsafehaven.ac.uk.                                                          
Verified that domain green.develop.turingsafehaven.ac.uk is delegated to Azure. 
```

After that, the SHM is deployed and you can move on to creating an SRE configuration and then deploying the SRE.

By the way, I think that when you send responses and output by email, much of the content is being censored by your institute, because all I'm seeing is ***@***.***, not any of the output.

If you intend to paste some output from the console into your replies, I'd suggest doing it directly on Github instead if at all possible.

**Comment by mattwestby** [comment]
Hi Matt,

Ah ok thanks, I wasn’t sure whether there was more for the SHM.

Could you send me an example sre config as I’m having problems with the validation of what I’m entering.

Thanks

Matt

**Comment by mattwestby** [comment]
I've got to the point now where im starting to deploy the sre but i now get this error;

![sre-deploy-error](https://github.com/user-attachments/assets/ecd7d9fa-2672-44f4-b96a-532f0b792d0b)

**Comment by JimMadge** [comment]
Closes as covered by #2197
